File: sdk/python/qubit_sdk.py
"""
Qubit Protocol Python SDK

Provides a high-level interface for interacting with the Qubit blockchain:
- Wallet management (key generation, signing)
- Transaction creation and broadcasting
- Block and transaction queries
- ZK-SNARK proof generation for private transactions
- VDF verification
- Neural Guardian threat detection queries
"""

import hashlib
import json
import time
from typing import Optional, Dict, List, Tuple
import requests
from dataclasses import dataclass, asdict


@dataclass
class Transaction:
    """Qubit transaction structure"""
    sender: str  # 64-char hex address (32 bytes)
    recipient: str  # 64-char hex address
    amount: int  # Amount in satoshis (1 QBT = 10^8 sats)
    fee: int  # Transaction fee
    nonce: int  # Sender nonce to prevent replay attacks
    timestamp: int  # Unix timestamp
    signature: Optional[str] = None  # 128-char hex signature
    zk_proof: Optional[str] = None  # Optional ZK-SNARK proof for privacy

    def to_dict(self) -> dict:
        return asdict(self)

    def serialize(self) -> str:
        """Serialize transaction for signing"""
        return json.dumps({
            'sender': self.sender,
            'recipient': self.recipient,
            'amount': self.amount,
            'fee': self.fee,
            'nonce': self.nonce,
            'timestamp': self.timestamp
        }, sort_keys=True)

    def hash(self) -> str:
        """Compute transaction hash (double SHA-256)"""
        first_hash = hashlib.sha256(self.serialize().encode()).digest()
        second_hash = hashlib.sha256(first_hash).hexdigest()
        return second_hash


@dataclass
class Block:
    """Qubit block structure"""
    index: int
    timestamp: int
    transactions: List[Dict]
    previous_hash: str
    merkle_root: str
    nonce: int
    difficulty: int
    vdf_output: Optional[str] = None
    vdf_proof: Optional[str] = None
    hash: Optional[str] = None

    @classmethod
    def from_dict(cls, data: dict) -> 'Block':
        return cls(**data)


class Wallet:
    """Qubit wallet for key management and signing"""
    
    def __init__(self, private_key: Optional[str] = None):
        """
        Initialize wallet with private key or generate new one
        
        Args:
            private_key: 64-char hex private key (32 bytes), or None to generate
        """
        if private_key:
            self.private_key = private_key
        else:
            self.private_key = self._generate_private_key()
        
        self.public_key = self._derive_public_key(self.private_key)
        self.address = self._derive_address(self.public_key)
    
    @staticmethod
    def _generate_private_key() -> str:
        """Generate random 256-bit private key"""
        import secrets
        return secrets.token_hex(32)
    
    @staticmethod
    def _derive_public_key(private_key: str) -> str:
        """Derive public key from private key using Ed25519"""
        # Simplified: In production, use proper Ed25519 key derivation
        data = bytes.fromhex(private_key)
        pub = hashlib.sha256(data + b"public").hexdigest()
        return pub
    
    @staticmethod
    def _derive_address(public_key: str) -> str:
        """Derive address from public key (SHA-256 hash)"""
        return hashlib.sha256(bytes.fromhex(public_key)).hexdigest()
    
    def sign(self, message: str) -> str:
        """
        Sign a message with the wallet's private key
        
        Args:
            message: Message to sign (typically serialized transaction)
            
        Returns:
            128-char hex signature (64 bytes)
        """
        # Simplified Ed25519 signature (production: use proper Ed25519)
        msg_hash = hashlib.sha256(message.encode()).digest()
        key_bytes = bytes.fromhex(self.private_key)
        sig_data = hashlib.sha256(key_bytes + msg_hash).digest()
        # Pad to 64 bytes
        signature = (sig_data + sig_data).hex()
        return signature
    
    @staticmethod
    def verify(message: str, signature: str, public_key: str) -> bool:
        """
        Verify a signature
        
        Args:
            message: Original message
            signature: 128-char hex signature
            public_key: 64-char hex public key
            
        Returns:
            True if signature is valid
        """
        # Simplified verification (production: use proper Ed25519)
        msg_hash = hashlib.sha256(message.encode()).digest()
        expected_sig = hashlib.sha256(bytes.fromhex(public_key) + msg_hash).digest()
        actual_sig = bytes.fromhex(signature)[:32]
        return expected_sig == actual_sig


class QubitClient:
    """Client for interacting with Qubit node RPC API"""
    
    def __init__(self, node_url: str = "http://localhost:8332"):
        """
        Initialize client
        
        Args:
            node_url: URL of Qubit node RPC endpoint
        """
        self.node_url = node_url.rstrip('/')
        self.session = requests.Session()
    
    def _rpc_call(self, method: str, params: dict = None) -> dict:
        """Make RPC call to node"""
        payload = {
            "jsonrpc": "2.0",
            "id": int(time.time() * 1000),
            "method": method,
            "params": params or {}
        }
        
        try:
            response = self.session.post(f"{self.node_url}/rpc", json=payload, timeout=10)
            response.raise_for_status()
            result = response.json()
            
            if "error" in result and result["error"]:
                raise Exception(f"RPC error: {result['error']}")
            
            return result.get("result", {})
        
        except requests.exceptions.RequestException as e:
            raise Exception(f"Network error: {e}")
    
    def get_balance(self, address: str) -> int:
        """
        Get balance for an address
        
        Args:
            address: 64-char hex address
            
        Returns:
            Balance in satoshis
        """
        return self._rpc_call("get_balance", {"address": address})
    
    def get_nonce(self, address: str) -> int:
        """
        Get current nonce for an address
        
        Args:
            address: 64-char hex address
            
        Returns:
            Current nonce
        """
        return self._rpc_call("get_nonce", {"address": address})
    
    def broadcast_transaction(self, tx: Transaction) -> str:
        """
        Broadcast a signed transaction
        
        Args:
            tx: Signed transaction
            
        Returns:
            Transaction hash
        """
        return self._rpc_call("broadcast_transaction", tx.to_dict())
    
    def get_transaction(self, tx_hash: str) -> Optional[Dict]:
        """
        Get transaction by hash
        
        Args:
            tx_hash: 64-char hex transaction hash
            
        Returns:
            Transaction data or None if not found
        """
        try:
            return self._rpc_call("get_transaction", {"hash": tx_hash})
        except:
            return None
    
    def get_block(self, block_hash: Optional[str] = None, index: Optional[int] = None) -> Optional[Block]:
        """
        Get block by hash or index
        
        Args:
            block_hash: 64-char hex block hash
            index: Block index
            
        Returns:
            Block object or None if not found
        """
        params = {}
        if block_hash:
            params["hash"] = block_hash
        elif index is not None:
            params["index"] = index
        else:
            raise ValueError("Must provide either block_hash or index")
        
        try:
            data = self._rpc_call("get_block", params)
            return Block.from_dict(data)
        except:
            return None
    
    def get_latest_block(self) -> Block:
        """Get the latest block in the chain"""
        data = self._rpc_call("get_latest_block")
        return Block.from_dict(data)
    
    def get_chain_info(self) -> Dict:
        """
        Get blockchain info
        
        Returns:
            Dict with height, difficulty, total_supply, etc.
        """
        return self._rpc_call("get_chain_info")
    
    def create_transaction(self, wallet: Wallet, recipient: str, amount: int, 
                          fee: int = 1000, use_zk: bool = False) -> Transaction:
        """
        Create and sign a transaction
        
        Args:
            wallet: Sender wallet
            recipient: Recipient address (64-char hex)
            amount: Amount in satoshis
            fee: Transaction fee (default 1000 sats = 0.00001 QBT)
            use_zk: Whether to generate ZK-SNARK proof for privacy
            
        Returns:
            Signed transaction ready to broadcast
        """
        nonce = self.get_nonce(wallet.address)
        
        tx = Transaction(
            sender=wallet.address,
            recipient=recipient,
            amount=amount,
            fee=fee,
            nonce=nonce,
            timestamp=int(time.time())
        )
        
        # Sign transaction
        tx.signature = wallet.sign(tx.serialize())
        
        # Generate ZK proof if requested
        if use_zk:
            tx.zk_proof = self._generate_zk_proof(tx, wallet)
        
        return tx
    
    def _generate_zk_proof(self, tx: Transaction, wallet: Wallet) -> str:
        """
        Generate ZK-SNARK proof for private transaction
        
        This proves the sender has sufficient balance without revealing
        the exact balance or transaction amount
        """
        # Call ZK prover service
        proof_data = self._rpc_call("generate_zk_proof", {
            "sender": tx.sender,
            "amount": tx.amount,
            "private_key": wallet.private_key  # Sent over secure channel
        })
        return proof_data.get("proof", "")
    
    def verify_vdf(self, vdf_output: str, vdf_proof: str, input_data: str, time_param: int) -> bool:
        """
        Verify VDF proof
        
        Args:
            vdf_output: VDF output value
            vdf_proof: Wesolowski proof
            input_data: VDF input
            time_param: Time parameter
            
        Returns:
            True if proof is valid
        """
        result = self._rpc_call("verify_vdf", {
            "output": vdf_output,
            "proof": vdf_proof,
            "input": input_data,
            "time": time_param
        })
        return result.get("valid", False)
    
    def query_neural_guardian(self, peer_id: str) -> Dict:
        """
        Query Neural Guardian for threat assessment of a peer
        
        Args:
            peer_id: Peer identifier
            
        Returns:
            Dict with threat_type, confidence, action
        """
        return self._rpc_call("neural_guardian_query", {"peer_id": peer_id})
    
    def send(self, wallet: Wallet, recipient: str, amount: int, 
            fee: int = 1000, use_zk: bool = False) -> str:
        """
        Convenience method to create, sign, and broadcast a transaction
        
        Args:
            wallet: Sender wallet
            recipient: Recipient address
            amount: Amount in satoshis
            fee: Transaction fee
            use_zk: Use ZK-SNARK for privacy
            
        Returns:
            Transaction hash
        """
        tx = self.create_transaction(wallet, recipient, amount, fee, use_zk)
        return self.broadcast_transaction(tx)


def qbt_to_sats(qbt: float) -> int:
    """Convert QBT to satoshis (1 QBT = 10^8 sats)"""
    return int(qbt * 100_000_000)


def sats_to_qbt(sats: int) -> float:
    """Convert satoshis to QBT"""
    return sats / 100_000_000


# Example usage
if __name__ == "__main__":
    # Initialize client
    client = QubitClient("http://localhost:8332")
    
    # Create or load wallet
    wallet = Wallet()  # Generate new wallet
    print(f"Address: {wallet.address}")
    
    # Check balance
    balance = client.get_balance(wallet.address)
    print(f"Balance: {sats_to_qbt(balance)} QBT")
    
    # Send transaction
    recipient = "a" * 64  # Example recipient address
    amount = qbt_to_sats(1.5)  # Send 1.5 QBT
    
    try:
        tx_hash = client.send(wallet, recipient, amount, use_zk=True)
        print(f"Transaction sent: {tx_hash}")
    except Exception as e:
        print(f"Error: {e}")
    
    # Get chain info
    info = client.get_chain_info()
    print(f"Chain height: {info.get('height')}")
    print(f"Total supply: {sats_to_qbt(info.get('total_supply', 0))} QBT")

File: sdk/javascript/qubit-sdk.js
/**
 * Qubit Protocol JavaScript SDK
 * 
 * Provides a high-level interface for interacting with the Qubit blockchain:
 * - Wallet management (key generation, signing)
 * - Transaction creation and broadcasting
 * - Block and transaction queries
 * - ZK-SNARK proof generation for private transactions
 * - VDF verification
 * - Neural Guardian threat detection queries
 */

const crypto = require('crypto');
const fetch = require('node-fetch');


/**
 * Transaction class
 */
class Transaction {
    constructor(sender, recipient, amount, fee, nonce, timestamp) {
        this.sender = sender;
        this.recipient = recipient;
        this.amount = amount;
        this.fee = fee;
        this.nonce = nonce;
        this.timestamp = timestamp;
        this.signature = null;
        this.zkProof = null;
    }

    /**
     * Serialize transaction for signing
     */
    serialize() {
        return JSON.stringify({
            sender: this.sender,
            recipient: this.recipient,
            amount: this.amount,
            fee: this.fee,
            nonce: this.nonce,
            timestamp: this.timestamp
        }, Object.keys({sender:0, recipient:0, amount:0, fee:0, nonce:0, timestamp:0}).sort());
    }

    /**
     * Compute transaction hash (double SHA-256)
     */
    hash() {
        const firstHash = crypto.createHash('sha256').update(this.serialize()).digest();
        const secondHash = crypto.createHash('sha256').update(firstHash).digest('hex');
        return secondHash;
    }

    /**
     * Convert to plain object
     */
    toJSON() {
        return {
            sender: this.sender,
            recipient: this.recipient,
            amount: this.amount,
            fee: this.fee,
            nonce: this.nonce,
            timestamp: this.timestamp,
            signature: this.signature,
            zk_proof: this.zkProof
        };
    }
}


/**
 * Block class
 */
class Block {
    constructor(data) {
        this.index = data.index;
        this.timestamp = data.timestamp;
        this.transactions = data.transactions;
        this.previousHash = data.previous_hash || data.previousHash;
        this.merkleRoot = data.merkle_root || data.merkleRoot;
        this.nonce = data.nonce;
        this.difficulty = data.difficulty;
        this.vdfOutput = data.vdf_output || data.vdfOutput;
        this.vdfProof = data.vdf_proof || data.vdfProof;
        this.hash = data.hash;
    }
}


/**
 * Wallet class for key management and signing
 */
class Wallet {
    /**
     * Initialize wallet with private key or generate new one
     * @param {string|null} privateKey - 64-char hex private key (32 bytes), or null to generate
     */
    constructor(privateKey = null) {
        if (privateKey) {
            this.privateKey = privateKey;
        } else {
            this.privateKey = this._generatePrivateKey();
        }
        
        this.publicKey = this._derivePublicKey(this.privateKey);
        this.address = this._deriveAddress(this.publicKey);
    }

    /**
     * Generate random 256-bit private key
     */
    _generatePrivateKey() {
        return crypto.randomBytes(32).toString('hex');
    }

    /**
     * Derive public key from private key using Ed25519
     */
    _derivePublicKey(privateKey) {
        // Simplified: In production, use proper Ed25519 key derivation
        const data = Buffer.from(privateKey, 'hex');
        const pub = crypto.createHash('sha256')
            .update(Buffer.concat([data, Buffer.from('public')]))
            .digest('hex');
        return pub;
    }

    /**
     * Derive address from public key (SHA-256 hash)
     */
    _deriveAddress(publicKey) {
        return crypto.createHash('sha256')
            .update(Buffer.from(publicKey, 'hex'))
            .digest('hex');
    }

    /**
     * Sign a message with the wallet's private key
     * @param {string} message - Message to sign (typically serialized transaction)
     * @returns {string} 128-char hex signature (64 bytes)
     */
    sign(message) {
        // Simplified Ed25519 signature (production: use proper Ed25519)
        const msgHash = crypto.createHash('sha256').update(message).digest();
        const keyBytes = Buffer.from(this.privateKey, 'hex');
        const sigData = crypto.createHash('sha256')
            .update(Buffer.concat([keyBytes, msgHash]))
            .digest();
        // Pad to 64 bytes
        const signature = Buffer.concat([sigData, sigData]).toString('hex');
        return signature;
    }

    /**
     * Verify a signature
     * @param {string} message - Original message
     * @param {string} signature - 128-char hex signature
     * @param {string} publicKey - 64-char hex public key
     * @returns {boolean} True if signature is valid
     */
    static verify(message, signature, publicKey) {
        // Simplified verification (production: use proper Ed25519)
        const msgHash = crypto.createHash('sha256').update(message).digest();
        const expectedSig = crypto.createHash('sha256')
            .update(Buffer.concat([Buffer.from(publicKey, 'hex'), msgHash]))
            .digest();
        const actualSig = Buffer.from(signature, 'hex').slice(0, 32);
        return expectedSig.equals(actualSig);
    }
}


/**
 * Client for interacting with Qubit node RPC API
 */
class QubitClient {
    /**
     * Initialize client
     * @param {string} nodeUrl - URL of Qubit node RPC endpoint
     */
    constructor(nodeUrl = 'http://localhost:8332') {
        this.nodeUrl = nodeUrl.replace(/\/$/, '');
    }

    /**
     * Make RPC call to node
     */
    async _rpcCall(method, params = {}) {
        const payload = {
            jsonrpc: '2.0',
            id: Date.now(),
            method: method,
            params: params
        };

        try {
            const response = await fetch(`${this.nodeUrl}/rpc`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(payload),
                timeout: 10000
            });

            if (!response.ok) {
                throw new Error(`HTTP error: ${response.status}`);
            }

            const result = await response.json();

            if (result.error) {
                throw new Error(`RPC error: ${JSON.stringify(result.error)}`);
            }

            return result.result || {};

        } catch (error) {
            throw new Error(`Network error: ${error.message}`);
        }
    }

    /**
     * Get balance for an address
     * @param {string} address - 64-char hex address
     * @returns {number} Balance in satoshis
     */
    async getBalance(address) {
        return await this._rpcCall('get_balance', { address });
    }

    /**
     * Get current nonce for an address
     * @param {string} address - 64-char hex address
     * @returns {number} Current nonce
     */
    async getNonce(address) {
        return await this._rpcCall('get_nonce', { address });
    }

    /**
     * Broadcast a signed transaction
     * @param {Transaction} tx - Signed transaction
     * @returns {string} Transaction hash
     */
    async broadcastTransaction(tx) {
        return await this._rpcCall('broadcast_transaction', tx.toJSON());
    }

    /**
     * Get transaction by hash
     * @param {string} txHash - 64-char hex transaction hash
     * @returns {object|null} Transaction data or null if not found
     */
    async getTransaction(txHash) {
        try {
            return await this._rpcCall('get_transaction', { hash: txHash });
        } catch {
            return null;
        }
    }

    /**
     * Get block by hash or index
     * @param {string|null} blockHash - 64-char hex block hash
     * @param {number|null} index - Block index
     * @returns {Block|null} Block object or null if not found
     */
    async getBlock(blockHash = null, index = null) {
        const params = {};
        if (blockHash) {
            params.hash = blockHash;
        } else if (index !== null) {
            params.index = index;
        } else {
            throw new Error('Must provide either blockHash or index');
        }

        try {
            const data = await this._rpcCall('get_block', params);
            return new Block(data);
        } catch {
            return null;
        }
    }

    /**
     * Get the latest block in the chain
     * @returns {Block} Latest block
     */
    async getLatestBlock() {
        const data = await this._rpcCall('get_latest_block');
        return new Block(data);
    }

    /**
     * Get blockchain info
     * @returns {object} Dict with height, difficulty, total_supply, etc.
     */
    async getChainInfo() {
        return await this._rpcCall('get_chain_info');
    }

    /**
     * Create and sign a transaction
     * @param {Wallet} wallet - Sender wallet
     * @param {string} recipient - Recipient address (64-char hex)
     * @param {number} amount - Amount in satoshis
     * @param {number} fee - Transaction fee (default 1000 sats = 0.00001 QBT)
     * @param {boolean} useZk - Whether to generate ZK-SNARK proof for privacy
     * @returns {Transaction} Signed transaction ready to broadcast
     */
    async createTransaction(wallet, recipient, amount, fee = 1000, useZk = false) {
        const nonce = await this.getNonce(wallet.address);
        
        const tx = new Transaction(
            wallet.address,
            recipient,
            amount,
            fee,
            nonce,
            Math.floor(Date.now() / 1000)
        );

        // Sign transaction
        tx.signature = wallet.sign(tx.serialize());

        // Generate ZK proof if requested
        if (useZk) {
            tx.zkProof = await this._generateZkProof(tx, wallet);
        }

        return tx;
    }

    /**
     * Generate ZK-SNARK proof for private transaction
     */
    async _generateZkProof(tx, wallet) {
        const proofData = await this._rpcCall('generate_zk_proof', {
            sender: tx.sender,
            amount: tx.amount,
            private_key: wallet.privateKey  // Sent over secure channel
        });
        return proofData.proof || '';
    }

    /**
     * Verify VDF proof
     * @param {string} vdfOutput - VDF output value
     * @param {string} vdfProof - Wesolowski proof
     * @param {string} inputData - VDF input
     * @param {number} timeParam - Time parameter
     * @returns {boolean} True if proof is valid
     */
    async verifyVdf(vdfOutput, vdfProof, inputData, timeParam) {
        const result = await this._rpcCall('verify_vdf', {
            output: vdfOutput,
            proof: vdfProof,
            input: inputData,
            time: timeParam
        });
        return result.valid || false;
    }

    /**
     * Query Neural Guardian for threat assessment of a peer
     * @param {string} peerId - Peer identifier
     * @returns {object} Dict with threat_type, confidence, action
     */
    async queryNeuralGuardian(peerId) {
        return await this._rpcCall('neural_guardian_query', { peer_id: peerId });
    }

    /**
     * Convenience method to create, sign, and broadcast a transaction
     * @param {Wallet} wallet - Sender wallet
     * @param {string} recipient - Recipient address
     * @param {number} amount - Amount in satoshis
     * @param {number} fee - Transaction fee
     * @param {boolean} useZk - Use ZK-SNARK for privacy
     * @returns {string} Transaction hash
     */
    async send(wallet, recipient, amount, fee = 1000, useZk = false) {
        const tx = await this.createTransaction(wallet, recipient, amount, fee, useZk);
        return await this.broadcastTransaction(tx);
    }
}


/**
 * Convert QBT to satoshis (1 QBT = 10^8 sats)
 */
function qbtToSats(qbt) {
    return Math.floor(qbt * 100_000_000);
}


/**
 * Convert satoshis to QBT
 */
function satsToQbt(sats) {
    return sats / 100_000_000;
}


// Export classes and functions
module.exports = {
    Transaction,
    Block,
    Wallet,
    QubitClient,
    qbtToSats,
    satsToQbt
};


// Example usage
if (require.main === module) {
    (async () => {
        // Initialize client
        const client = new QubitClient('http://localhost:8332');
        
        // Create or load wallet
        const wallet = new Wallet();  // Generate new wallet
        console.log(`Address: ${wallet.address}`);
        
        try {
            // Check balance
            const balance = await client.getBalance(wallet.address);
            console.log(`Balance: ${satsToQbt(balance)} QBT`);
            
            // Send transaction
            const recipient = 'a'.repeat(64);  // Example recipient address
            const amount = qbtToSats(1.5);  // Send 1.5 QBT
            
            const txHash = await client.send(wallet, recipient, amount, 1000, true);
            console.log(`Transaction sent: ${txHash}`);
            
            // Get chain info
            const info = await client.getChainInfo();
            console.log(`Chain height: ${info.height}`);
            console.log(`Total supply: ${satsToQbt(info.total_supply || 0)} QBT`);
            
        } catch (error) {
            console.error(`Error: ${error.message}`);
        }
    })();
}
